### ISSUE TITLE
TargetEncoder [FEATURE]

### ISSUE THREAD

**Issue by AlanGanem** [koaning/scikit-lego#427]
# **Please explain clearly what you'd like to see added.** 

When modelling problems with high cardinality categorical features, one possible way to encode them is through target enconding. It prevents sparsity of one hot encoding. Target enconding basically encodes the feature labels based on some target variable dependent metric (aggregate by feature label and perform a groupwise calculation), on the trainning set, saves a lookup table in the estiamtor state and performs a simple lookup during transform. It works for both classification and regression problems.

I already have this implemented in a sklearn fashion (fit-transform), and though i'm not inheriting from sklearn abstract classes, it porbably won't be hard to fit it into the requirements.

### Considering the Iris Dataset, i've created some random categorical variables **cat_col1** and **cat_col2**:
![image](https://user-images.githubusercontent.com/47705293/102896706-91886780-4445-11eb-9ff5-ecd84da6df49.png)

### Now lets encode **cat_col1** and **cat_col2** considering a classification problem with dependet variable **target** and [Weight of Evidence](https://www.kaggle.com/davidbnn92/weight-of-evidence-encoding) as aggregation metric. Here's the output:
![image](https://user-images.githubusercontent.com/47705293/102896796-b7ae0780-4445-11eb-837a-39c448f482ad.png)

### Now let's consider we want to predict **petal width** (a regression problem). We can do the same thing, using the mean as aggregation function (you can stack multiple encoders such as mean encoder, variance encoder, nth_percentile...):

![image](https://user-images.githubusercontent.com/47705293/102896993-16738100-4446-11eb-94d6-56caf850ae41.png)

- [ ] convince us of the use-case, we're open to many suggestions but we prefer to solve problems with pipelines that are at least somewhat general 
- [X] add a screenshot if applicable (ML stuff is hard to explain with words, pictures say 1000 words)
- [x] make sure that the feature you want is not already supported by sklearn

**Comment by koaning** [comment]
Just to check, isn't what you're referring to already implemented [here](https://dirty-cat.github.io/stable/generated/dirty_cat.TargetEncoder.html#dirty_cat.TargetEncoder).

**Comment by AlanGanem** [comment]
Yes, that's perfect. Thanks for sharing. 
Still i'm not too sure if this implementation takes into account out-of-the-group information when calculating the metric, like in the WOE case, since its calculated as:

$WOE =  \log{\frac{(\frac{gp}{tp})}{(\frac{gn}{tn})}}$

where `tp` is the total positives for that class in the entire dataset and `tn` is the total negatives for that class in the entire dataset.

But i do agree this would be more of a dirty_cat concern rather than sklego's

**Comment by koaning** [comment]
Alternatively there also [cateogry encoders](https://github.com/scikit-learn-contrib/category_encoders). Given the number of packages where this feature feels more appropriate, I'll this the issue here.
